Sam 2 Auto Annotator Error

### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.


### Ultralytics YOLO Component

Predict

### Bug

Getting an ```AssertionError``` after first image is auto annotated using ```auto_annotate```  tool :

```
from ultralytics.data.annotator import auto_annotate

auto_annotate(data="videos/bedroom", det_model="yolov8x.pt", sam_model="sam2_b.pt")
```
Got this output:

`image 1/200 /home/hmehdi/segment-anything-2/notebooks/videos/bedroom/00000.jpg: 384x640 2 persons, 1 bed, 22.4ms`

Then this error:
```
{
	"name": "AssertionError",
	"message": "",
	"stack": "---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[4], line 3
      1 from ultralytics.data.annotator import auto_annotate
----> 3 auto_annotate(data=\"videos/bedroom\", det_model=\"yolov8x.pt\", sam_model=\"sam2_b.pt\")

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/data/annotator.py:45, in auto_annotate(data, det_model, sam_model, device, output_dir)
     43 if len(class_ids):
     44     boxes = result.boxes.xyxy  # Boxes object for bbox outputs
---> 45     sam_results = sam_model(result.orig_img, bboxes=boxes, verbose=False, save=False, device=device)
     46     segments = sam_results[0].masks.xyn  # noqa
     48     with open(f\"{Path(output_dir) / Path(result.path).stem}.txt\", \"w\") as f:

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam/model.py:98, in SAM.__call__(self, source, stream, bboxes, points, labels, **kwargs)
     84 def __call__(self, source=None, stream=False, bboxes=None, points=None, labels=None, **kwargs):
     85     \"\"\"
     86     Alias for the 'predict' method.
     87 
   (...)
     96         (list): The model predictions.
     97     \"\"\"
---> 98     return self.predict(source, stream, bboxes, points, labels, **kwargs)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam/model.py:82, in SAM.predict(self, source, stream, bboxes, points, labels, **kwargs)
     80 kwargs.update(overrides)
     81 prompts = dict(bboxes=bboxes, points=points, labels=labels)
---> 82 return super().predict(source, stream, prompts=prompts, **kwargs)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/engine/model.py:563, in Model.predict(self, source, stream, predictor, **kwargs)
    561 if prompts and hasattr(self.predictor, \"set_prompts\"):  # for SAM-type models
    562     self.predictor.set_prompts(prompts)
--> 563 return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/engine/predictor.py:168, in BasePredictor.__call__(self, source, model, stream, *args, **kwargs)
    166     return self.stream_inference(source, model, *args, **kwargs)
    167 else:
--> 168     return list(self.stream_inference(source, model, *args, **kwargs))

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/torch/utils/_contextlib.py:36, in _wrap_generator.<locals>.generator_context(*args, **kwargs)
     33 try:
     34     # Issuing `None` to a generator fires it up
     35     with ctx_factory():
---> 36         response = gen.send(None)
     38     while True:
     39         try:
     40             # Forward the response to our caller and get its next request

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/engine/predictor.py:254, in BasePredictor.stream_inference(self, source, model, *args, **kwargs)
    252 # Inference
    253 with profilers[1]:
--> 254     preds = self.inference(im, *args, **kwargs)
    255     if self.args.embed:
    256         yield from [preds] if isinstance(preds, torch.Tensor) else preds  # yield embedding tensors

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam/predict.py:150, in Predictor.inference(self, im, bboxes, points, labels, masks, multimask_output, *args, **kwargs)
    147 if all(i is None for i in [bboxes, points, masks]):
    148     return self.generate(im, *args, **kwargs)
--> 150 return self.prompt_inference(im, bboxes, points, labels, masks, multimask_output)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam2/predict.py:132, in SAM2Predictor.prompt_inference(self, im, bboxes, points, labels, masks, multimask_output, img_idx)
    130 batched_mode = points is not None and points[0].shape[0] > 1  # multi object prediction
    131 high_res_features = [feat_level[img_idx].unsqueeze(0) for feat_level in features[\"high_res_feats\"]]
--> 132 pred_masks, pred_scores, _, _ = self.model.sam_mask_decoder(
    133     image_embeddings=features[\"image_embed\"][img_idx].unsqueeze(0),
    134     image_pe=self.model.sam_prompt_encoder.get_dense_pe(),
    135     sparse_prompt_embeddings=sparse_embeddings,
    136     dense_prompt_embeddings=dense_embeddings,
    137     multimask_output=multimask_output,
    138     repeat_image=batched_mode,
    139     high_res_features=high_res_features,
    140 )
    141 # (N, d, H, W) --> (N*d, H, W), (N, d) --> (N*d, )
    142 # `d` could be 1 or 3 depends on `multimask_output`.
    143 return pred_masks.flatten(0, 1), pred_scores.flatten(0, 1)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/torch/nn/modules/module.py:1553, in Module._wrapped_call_impl(self, *args, **kwargs)
   1551     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1552 else:
-> 1553     return self._call_impl(*args, **kwargs)

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/torch/nn/modules/module.py:1562, in Module._call_impl(self, *args, **kwargs)
   1557 # If we don't have any hooks, we want to skip the rest of the logic in
   1558 # this function, and just call forward.
   1559 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1560         or _global_backward_pre_hooks or _global_backward_hooks
   1561         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1562     return forward_call(*args, **kwargs)
   1564 try:
   1565     result = None

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam2/modules/decoders.py:156, in MaskDecoder.forward(self, image_embeddings, image_pe, sparse_prompt_embeddings, dense_prompt_embeddings, multimask_output, repeat_image, high_res_features)
    119 def forward(
    120     self,
    121     image_embeddings: torch.Tensor,
   (...)
    127     high_res_features: Optional[List[torch.Tensor]] = None,
    128 ) -> Tuple[torch.Tensor, torch.Tensor]:
    129     \"\"\"
    130     Predicts masks given image and prompt embeddings.
    131 
   (...)
    154         ...     sparse_prompt_embeddings, dense_prompt_embeddings, True, False)
    155     \"\"\"
--> 156     masks, iou_pred, mask_tokens_out, object_score_logits = self.predict_masks(
    157         image_embeddings=image_embeddings,
    158         image_pe=image_pe,
    159         sparse_prompt_embeddings=sparse_prompt_embeddings,
    160         dense_prompt_embeddings=dense_prompt_embeddings,
    161         repeat_image=repeat_image,
    162         high_res_features=high_res_features,
    163     )
    165     # Select the correct mask or masks for output
    166     if multimask_output:

File ~/anaconda3/envs/sam2/lib/python3.12/site-packages/ultralytics/models/sam2/modules/decoders.py:219, in MaskDecoder.predict_masks(self, image_embeddings, image_pe, sparse_prompt_embeddings, dense_prompt_embeddings, repeat_image, high_res_features)
    217     src = torch.repeat_interleave(image_embeddings, tokens.shape[0], dim=0)
    218 else:
--> 219     assert image_embeddings.shape[0] == tokens.shape[0]
    220     src = image_embeddings
    221 src = src + dense_prompt_embeddings

AssertionError: "
}
```

### Environment

```
Ultralytics YOLOv8.2.70  Python-3.12.3 torch-2.4.0+cu121 CUDA:0 (NVIDIA A100-SXM4-80GB, 81158MiB)
Setup complete ✅ (256 CPUs, 503.4 GB RAM, 4162.9/6444.6 GB disk)

OS                  Linux-5.15.0-117-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.12.3
Install             pip
RAM                 503.43 GB
CPU                 AMD EPYC 7713 64-Core Processor
CUDA                12.1

numpy               ✅ 1.26.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.1>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.4.0>=1.8.0
torchvision         ✅ 0.19.0>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0
```

### Minimal Reproducible Example

```
from ultralytics.data.annotator import auto_annotate

auto_annotate(data="videos/bedroom", det_model="yolov8x.pt", sam_model="sam2_b.pt")
```

videos/bedroom being a folder of frames that was provided from the official [segment-anything-2 github](https://github.com/facebookresearch/segment-anything-2/tree/main/notebooks/videos/bedroom)

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

--- a/ultralytics/__init__.py
+++ b/ultralytics/__init__.py
@@ -1,6 +1,6 @@
 # Ultralytics YOLO 🚀, AGPL-3.0 license
 
-__version__ = "8.2.71"
+__version__ = "8.2.72"
 
 import os
 

--- a/ultralytics/models/sam2/predict.py
+++ b/ultralytics/models/sam2/predict.py
@@ -102,23 +102,21 @@ class SAM2Predictor(Predictor):
         if bboxes is not None:
             bboxes = torch.as_tensor(bboxes, dtype=torch.float32, device=self.device)
             bboxes = bboxes[None] if bboxes.ndim == 1 else bboxes
-            bboxes *= r
+            bboxes = bboxes.view(-1, 2, 2) * r
+            bbox_labels = torch.tensor([[2, 3]], dtype=torch.int, device=bboxes.device)
+            bbox_labels = bbox_labels.repeat(bboxes.size(0), 1)
+            # NOTE: merge "boxes" and "points" into a single "points" input
+            # (where boxes are added at the beginning) to model.sam_prompt_encoder
+            if points is not None:
+                points = torch.cat([bboxes, points], dim=1)
+                labels = torch.cat([bbox_labels, labels], dim=1)
+                points = (points, labels)
+            else:
+                points = (bboxes, bbox_labels)
         if masks is not None:
             masks = torch.as_tensor(masks, dtype=torch.float32, device=self.device).unsqueeze(1)
 
         points = (points, labels) if points is not None else None
-        if bboxes is not None:
-            box_coords = bboxes.reshape(-1, 2, 2)
-            box_labels = torch.tensor([[2, 3]], dtype=torch.int, device=bboxes.device)
-            box_labels = box_labels.repeat(bboxes.size(0), 1)
-            # NOTE: merge "boxes" and "points" into a single "points" input 
-            # (where boxes are added at the beginning) to model.sam_prompt_encoder
-            if points is not None:
-                concat_coords = torch.cat([box_coords, points[0]], dim=1)
-                concat_labels = torch.cat([box_labels, points[1]], dim=1)
-                points = (concat_coords, concat_labels)
-            else:
-                points = (box_coords, box_labels)
 
         sparse_embeddings, dense_embeddings = self.model.sam_prompt_encoder(
             points=points,

--- a/ultralytics/models/sam2/predict.py
+++ b/ultralytics/models/sam2/predict.py
@@ -103,16 +103,14 @@ class SAM2Predictor(Predictor):
             bboxes = torch.as_tensor(bboxes, dtype=torch.float32, device=self.device)
             bboxes = bboxes[None] if bboxes.ndim == 1 else bboxes
             bboxes = bboxes.view(-1, 2, 2) * r
-            bbox_labels = torch.tensor([[2, 3]], dtype=torch.int, device=bboxes.device)
-            bbox_labels = bbox_labels.repeat(bboxes.size(0), 1)
+            bbox_labels = torch.tensor([[2, 3]], dtype=torch.int32, device=bboxes.device).expand(len(bboxes), -1)
             # NOTE: merge "boxes" and "points" into a single "points" input
             # (where boxes are added at the beginning) to model.sam_prompt_encoder
             if points is not None:
                 points = torch.cat([bboxes, points], dim=1)
                 labels = torch.cat([bbox_labels, labels], dim=1)
-                points = (points, labels)
             else:
-                points = (bboxes, bbox_labels)
+                points, labels = bboxes, bbox_labels
         if masks is not None:
             masks = torch.as_tensor(masks, dtype=torch.float32, device=self.device).unsqueeze(1)
 

--- a/ultralytics/models/sam2/predict.py
+++ b/ultralytics/models/sam2/predict.py
@@ -106,11 +106,8 @@ class SAM2Predictor(Predictor):
             bbox_labels = torch.tensor([[2, 3]], dtype=torch.int32, device=bboxes.device).expand(len(bboxes), -1)
             # NOTE: merge "boxes" and "points" into a single "points" input
             # (where boxes are added at the beginning) to model.sam_prompt_encoder
-            if points is not None:
-                points = torch.cat([bboxes, points], dim=1)
-                labels = torch.cat([bbox_labels, labels], dim=1)
-            else:
-                points, labels = bboxes, bbox_labels
+            points = torch.cat([bboxes, points], dim=1) if points is not None else bboxes
+            labels = torch.cat([bbox_labels, labels], dim=1) if points is not None else bbox_labels
         if masks is not None:
             masks = torch.as_tensor(masks, dtype=torch.float32, device=self.device).unsqueeze(1)
 

--- a/ultralytics/models/sam2/predict.py
+++ b/ultralytics/models/sam2/predict.py
@@ -106,8 +106,11 @@ class SAM2Predictor(Predictor):
             bbox_labels = torch.tensor([[2, 3]], dtype=torch.int32, device=bboxes.device).expand(len(bboxes), -1)
             # NOTE: merge "boxes" and "points" into a single "points" input
             # (where boxes are added at the beginning) to model.sam_prompt_encoder
-            points = torch.cat([bboxes, points], dim=1) if points is not None else bboxes
-            labels = torch.cat([bbox_labels, labels], dim=1) if points is not None else bbox_labels
+            if points is not None:
+                points = torch.cat([bboxes, points], dim=1)
+                labels = torch.cat([bbox_labels, labels], dim=1)
+            else:
+                points, labels = bboxes, bbox_labels
         if masks is not None:
             masks = torch.as_tensor(masks, dtype=torch.float32, device=self.device).unsqueeze(1)
 

--- a/ultralytics/models/sam2/predict.py
+++ b/ultralytics/models/sam2/predict.py
@@ -107,23 +107,22 @@ class SAM2Predictor(Predictor):
             masks = torch.as_tensor(masks, dtype=torch.float32, device=self.device).unsqueeze(1)
 
         points = (points, labels) if points is not None else None
-        # TODO: Embed prompts
-        # if bboxes is not None:
-        #     box_coords = bboxes.reshape(-1, 2, 2)
-        #     box_labels = torch.tensor([[2, 3]], dtype=torch.int, device=bboxes.device)
-        #     box_labels = box_labels.repeat(bboxes.size(0), 1)
-        #     # we merge "boxes" and "points" into a single "concat_points" input (where
-        #     # boxes are added at the beginning) to sam_prompt_encoder
-        #     if concat_points is not None:
-        #         concat_coords = torch.cat([box_coords, concat_points[0]], dim=1)
-        #         concat_labels = torch.cat([box_labels, concat_points[1]], dim=1)
-        #         concat_points = (concat_coords, concat_labels)
-        #     else:
-        #         concat_points = (box_coords, box_labels)
+        if bboxes is not None:
+            box_coords = bboxes.reshape(-1, 2, 2)
+            box_labels = torch.tensor([[2, 3]], dtype=torch.int, device=bboxes.device)
+            box_labels = box_labels.repeat(bboxes.size(0), 1)
+            # NOTE: merge "boxes" and "points" into a single "points" input 
+            # (where boxes are added at the beginning) to model.sam_prompt_encoder
+            if points is not None:
+                concat_coords = torch.cat([box_coords, points[0]], dim=1)
+                concat_labels = torch.cat([box_labels, points[1]], dim=1)
+                points = (concat_coords, concat_labels)
+            else:
+                points = (box_coords, box_labels)
 
         sparse_embeddings, dense_embeddings = self.model.sam_prompt_encoder(
             points=points,
-            boxes=bboxes,
+            boxes=None,
             masks=masks,
         )
         # Predict masks

